src/components/__tests__/change-view-test.jsx

```diff
@@ -0,0 +1,379 @@
+            const relevantTypes = diffTypesFor(mediaType);
+
+            const changeView = shallow(
+              <ChangeView
+                page={simplePage}
+                from={{content_type: mediaType}}
+                to={{content_type: mediaType}}
+                user={{email: 'me'}}
+              />,
+              {context: {api: mockApi}}
+            );
+
+            expect(changeView.state().diffType).toBe(relevantTypes[0].value);
+          });
+        });
+      });
+    });
+
+    describe('when a diffType has NOT been stored in layeredStorage', () => {
+      describe('when defaultDiffType (SIDE_BY_SIDE_RENDERED) is relevant to the pages being compared', () => {
+        it('sets state.diffType to SIDE_BY_SIDE_RENDERED', () => {
+          layeredStorage.getItem.mockReturnValue(null);
+
+          const changeView = shallow(
+            <ChangeView
+              page={simplePage}
+              from={{content_type: 'text/html'}}
+              to={{content_type: 'text/html'}}
+              user={{email: 'me'}}
+            />,
+            {context: {api: mockApi}}
+          );
+
+          expect(changeView.state().diffType).toBe('SIDE_BY_SIDE_RENDERED');
+        });
+      });
+
+      describe('when defaultDiffType (SIDE_BY_SIDE_RENDERED) is NOT relevant to the pages being compared', () => {
+        it('sets state.diffType to the first relevant diff type', () => {
+          layeredStorage.getItem.mockReturnValue(null);
+
+          const mediaType = 'text/*';
+          const relevantTypes = diffTypesFor(mediaType);
+
+          const changeView = shallow(
+            <ChangeView
+              page={simplePage}
+              from={{content_type: mediaType}}
+              to={{content_type: mediaType}}
+              user={{email: 'me'}}
+            />,
+            {context: {api: mockApi}}
+          );
+
+          expect(changeView.state().diffType).toBe(relevantTypes[0].value);
+        });
+      });
+    });
+  });
+
+  describe('when the page versions change via props', () => {
+    describe('when state.diffType is relevant to the new pages being compared', () => {
+      it('leaves state.diffType at its current value', () => {
+
+        const oldMediaType = 'text/html';
+        const newMediaType = 'text/*';
+
+        const newRelevantTypes = diffTypesFor(newMediaType);
+        const diffType = newRelevantTypes[0].value;
+
+        const changeView = shallow(
+          <ChangeView
+            page={simplePage}
+            from={{content_type: oldMediaType}}
+            to={{content_type: oldMediaType}}
+            user={{email: 'me'}}
+          />,
+          {context: {api: mockApi}}
+        );
+
+        changeView.setState({diffType});
+
+        changeView.setProps({
+          from: {content_type: newMediaType},
+          to: {content_type: newMediaType},
+        });
+
+        expect(changeView.state().diffType).toBe(diffType);
+      });
```

Comment:
I’m not sure I’m following what this test is supposed to be. If we’re testing that the state gets left as-is, why are we calling `changeView.setState({diffType})`? And then we are asserting that the diff type wound up being one that was relevant for the new types, instead of whatever would have been set for the original types.

Reply:
Same for the rest of these where we’re calling `setState()` from the outside.

Reply:
Yeah I agree this is a little weird :/ My thinking in using setState here (and subsequently) was to set up an initial state for the test, after the effects of `getDerivedStateFromProps` on initial render, which doesn't always set up necessary conditions for the test. For example, in this case the steps would go like this:

1. initial render with `text/html` files, state.diffType gets initialized as `defaultDiffType`

2. What we are trying to test here is that if the files passed in through props change, but the value in `state.diffType` is relevant to the files in the incoming props, state.diffType will remain the same. Since `defaultDiffType` is not relevant to the file type about to be passed through props (`text/xml`), we need to manually set state.diffType to a diffType that IS relevant (in this case, `HIGHLIGHGHTED_SOURCE`). Do this.

3. Pass different "to" and "from" props into the component with contentType `text/xml`. Make sure that `state.diffType` is preserved. This means that `state.diffType` is indeed prioritized over all other choices in `ensureValidDiffType`

Whew! hopefully that made a little sense. The rationale for subsequent tests is similar. If this is overkill and/or you have another approach, I'd be happy to simplify these!

Reply:
> Since `defaultDiffType` is not relevant to the file type about to be passed through props (`text/xml`), we need to manually set state.diffType to a diffType that IS relevant (in this case, `HIGHLIGHGHTED_SOURCE`). Do this.

Yeah, I guess this what I’m not getting. Why not just set the original two versions to ones that would have a diffType that would also be relevant to the new versions you are setting?

Reply:
Ok, I just looked at this again and you're totally right. setState isn't necessary to setup initial conditions for any of these tests. Removed it from all of them.